### ISSUE TITLE
Does there need to be a section on development?

### ISSUE THREAD

**Issue by tomdxw** [dxw/playbook#11]
Possibly split up into Rails and GovPress subsections?

We could do with specifying what coding style/standards we follow (i.e. for PHP we follow [PSR-2](http://www.php-fig.org/psr/psr-2/), for JS and SCSS/LESS we don't follow any particular style but we use 4 spaces to indent).

I've also been thinking about guidelines for writing plugins, I don't know if we want to go into that level of detail:

https://gist.github.com/tomdxw/9799113567ae135b9dfb

And we could write something similar about the structure of govsite-style themes too.

**Comment by harry-m** [comment]
We do need a section on development, and also on research. It should be here: 

https://github.com/dxw/playbook/blob/master/playbook.md#build

@onishiweb will be working on this section (with help from team no doubt) when time permits.

I think more detailed things like plugin guidelines and coding standards will go in guides though. See: https://github.com/dxw/playbook#guides

**Comment by dgmstuart** [comment]
I certainly haven't been following PSR-2 :P 

I feel quite strongly that coding standards are pointless unless automatically enforced, e.g. by linters and ideally CI.

Also that anything we write about development should be very lightweight and focus on QA - recognising that for most other things individual teams should do what makes sense for them.

**Comment by tomdxw** [comment]
> I feel quite strongly that coding standards are pointless unless automatically enforced, e.g. by linters and ideally CI.

That's a good point. It's quite a simple change to existing .travis.yml files: https://github.com/dxw/php-missing/commit/cffffe06709695bb5764f6c1e3fae10e28b6fa5f

This does raise one issue though - we have no tests being run automatically on private repos on the WordPress side of the fence. I know the Rails team have Jenkins set up.

I tried setting up Jenkins for our private repos before but it seems like a lot of work. Since we're already using Travis-CI on all our public repos ( https://travis-ci.org/dxw ), is there any chance we can use the paid version ( https://travis-ci.com/plans )? @harry-m ?

**Comment by harry-m** [comment]
@tomdxw I have questions about Travis.

For example: can we set it up to pull project repos and run tests in those? Can we do things like run grunt commands to make sure all the assets compile and things successfully lint? Is two concurrent jobs enough for us to run Travis on all the things we do?

Perhaps it would be better to say: what can we not do with hosted Travis?

@leedxw do you or Tamas have any thoughts about using hosted Travis rather than running a CI ourselves?

**Comment by tomdxw** [comment]
> For example: can we set it up to pull project repos and run tests in those?

That's pretty much what Travis does. When any branch of a repo is pushed to it clones it, checks out the relevant branch, looks at the .travis.yml file in the root of the repo, and executes whatever's in there.

> Can we do things like run grunt commands to make sure all the assets compile and things successfully lint?

Yes. Travis won't commit anything, but it can make sure that when you do run grunt that it works, and it can make sure that linters are happy.

See this commit where I add the php-cs-fixer linter to .travis.yml: https://github.com/dxw/php-missing/commit/cffffe06709695bb5764f6c1e3fae10e28b6fa5f

> Is two concurrent jobs enough for us to run Travis on all the things we do?

Good question. On php-missing we test 6 different versions of PHP which means 6 tests need to pass before we can merge into master, and they take less than a minute each. So that would take about 3 minutes with 2 concurrent jobs.

Considering we're trying to start doing code reviews, that should easily be finished before somebody's had a chance to read the contents of a PR. And we can probably further increase our speed by testing fewer versions of PHP (as far as I'm aware we're not planning on switching to 5.4 or nightly or hhvm any time soon).

> Perhaps it would be better to say: what can we not do with hosted Travis?

I don't know about the paid version, but with the free version you can't configure who gets emails and when.

**Comment by onishiweb** [comment]
I've made a start to this with PRs #14 and #19 (the latter still to be reviewed/merged)

What else do we need to do? Looks like we need some linters set up for front-end stuff, which I'm happy to work on for Sass based on [code style](https://github.com/dxw/playbook/blob/master/playbook.md#code-style)

@tomdxw I see you've been working on Travis there, is there anything I can do to help or anything you need from me to help?

Also, any feedback on the above PRs would be welcome.

**Comment by harry-m** [comment]
@tomdxw I think there's quite a lot in there on development now. Can you have a look and let me know if there's anything you think is missing that we should include?

**Comment by harry-m** [comment]
I think this is in a good state now. Closing.
